src/common/hyperbeam/hb.ts

```diff
@@ -0,0 +1,232 @@
+   *
+   * @returns The meta data for the process
+   *
+   * @example
+   * const hyperbeam = new Hyperbeam({ url: 'https://hyperbeam.ario.permaweb.services', processId: 'qNvAoz0TgcH7DMg8BCVn8jF32QH5L6T29VjHxhHqqGE' });
+   * const meta = await hyperbeam.meta();
+   * console.log(meta);
+   */
+  async meta(): Promise<Record<string, JSONValue>> {
+    const url = new URL(`${this.url}/${this.processId}~process@1.0/meta`);
+    return this.fetchHyperbeamPath<Record<string, JSONValue>>({
+      path: url.toString(),
+    });
+  }
+
+  /**
+   * calls the process device /now function, which evaluates the current process state pulling new messages
+   * to get the latest state
+   *
+   * @param path - The path to the hb state
+   * @param json - Whether to return the result as JSON, defaults to true
+   * @returns The result of the compute operation
+   *
+   * @example
+   * const hyperbeam = new Hyperbeam({ url: 'https://hyperbeam.ario.permaweb.services', processId: 'qNvAoz0TgcH7DMg8BCVn8jF32QH5L6T29VjHxhHqqGE' });
+   * const result = await hyperbeam.now({ path: 'balances/QGWqtJdLLgm2ehFWiiPzMaoFLD50CnGuzZIPEdoDRGQ' });
+   * console.log(result);
+   */
+  async now<T extends JSONValue>({
+    path,
+    json = false,
+  }: {
+    path: string;
+    json?: boolean;
+  }): Promise<T> {
+    return this.fetchHyperbeamPath<T>({
+      path: `${this.url}/${this.processId}~process@1.0/now/${path}`,
+      json,
+    });
+  }
+
+  /**
+   * calls the process device /compute function, which uses the currently evaluated state in the node
+   *
+   * @param path - The path to the compute resource
+   * @param json - Whether to return the result as JSON, defaults to true
+   * @returns The result of the compute operation
+   *
+   * @example
+   * const hyperbeam = new Hyperbeam({ url: 'https://hyperbeam.ario.permaweb.services', processId: 'qNvAoz0TgcH7DMg8BCVn8jF32QH5L6T29VjHxhHqqGE' });
+   * const result = await hyperbeam.compute({ path: 'balances/QGWqtJdLLgm2ehFWiiPzMaoFLD50CnGuzZIPEdoDRGQ' });
+   * console.log(result);
+   */
+  async compute<T extends JSONValue>({
+    path,
+    json = false,
+  }: {
+    path: string;
+    json?: boolean;
+  }): Promise<T> {
+    return this.fetchHyperbeamPath<T>({
+      path: `${this.url}/${this.processId}~process@1.0/compute/${path}`,
+      json,
+    });
+  }
+
+  /**
+   * Checks if the process is HyperBeam compatible and caches the result.
+   *
+   * @returns {Promise<boolean>} True if the process is HyperBeam compatible, false otherwise.
+   */
+  async checkHyperBeamCompatibility({
+    minSlot,
+  }: {
+    minSlot?: number;
+  } = {}): Promise<boolean> {
+    // refetch if min slot is provided
+    if (minSlot !== undefined) {
+      this.isHyperBeamCompatible = undefined;
+      this.checkHyperBeamPromise = undefined;
+    }
+    if (this.checkHyperBeamPromise !== undefined) {
+      return this.checkHyperBeamPromise;
+    }
+
+    if (this.isHyperBeamCompatible !== undefined) {
```

Comment:
It's not clear to me how this case would ever be hit since you're not removing the reference to the promise before it resolves. You effectively end up with the same logical outcome since you always resolve the same value as you would have cached here. My initial suggestion to you was to add a function `isHyperBeamAvailable(): Promise<boolean>` that simply returns the cached promise and to have `checkHyperBeamCompatibility` ALWAYS reassign the old promise reference.